File: scala/day19_part1_2017.scala

import scala.io.Source

object TubeRouting {

  def main(args: Array[String]): Unit = {
    val lines = Source.fromFile("input.txt").getLines().toArray
    val (startX, startY) = findStartPosition(lines)
    val (letters, _) = followPath(lines, startX, startY)
    println(letters.mkString(""))
  }

  def findStartPosition(lines: Array[String]): (Int, Int) = {
    val startX = lines(0).indexOf('|')
    (startX, 0)
  }

  def followPath(lines: Array[String], startX: Int, startY: Int): (List[Char], (Int, Int)) = {
    var x = startX
    var y = startY
    var direction = (0, 1) // Start moving down
    var letters = List[Char]()

    while (y < lines.length && y >= 0 && x < lines(y).length && x >= 0) {
      val currentChar = lines(y)(x)

      currentChar match {
        case '|' | '-' =>
          // Continue in the same direction
        case '+' =>
          // Change direction
          direction = turn(lines, x, y, direction)
        case c if c.isLetter =>
          letters = letters :+ c
        case ' ' =>
          // If we hit a space, we should stop
          return (letters, (x, y))
      }

      // Move in the current direction
      x += direction._1
      y += direction._2
    }

    (letters, (x, y))
  }

  def turn(lines: Array[String], x: Int, y: Int, direction: (Int, Int)): (Int, Int) = {
    val (dx, dy) = direction
    val possibleTurns = List(
      (dx, dy) match {
        case (0, 1) => List((1, 0), (-1, 0)) // Down can turn right or left
        case (1, 0) => List((0, 1), (0, -1)) // Right can turn down or up
        case (0, -1) => List((1, 0), (-1, 0)) // Up can turn right or left
        case (-1, 0) => List((0, 1), (0, -1)) // Left can turn down or up
      }
    ).flatten

    for ((newDx, newDy) <- possibleTurns) {
      val newX = x + newDx
      val newY = y + newDy
      if (newY >= 0 && newY < lines.length && newX >= 0 && newX < lines(newY).length && 
          (lines(newY)(newX) == '|' || lines(newY)(newX) == '-' || lines(newY)(newX).isLetter || lines(newY)(newX) == '+')) {
        return (newDx, newDy)
      }
    }

    direction // Should not reach here
  }
}

File: haskell/day19_part1_2017.hs
import System.IO

data Direction = Up | Down | Lft | Rght deriving (Eq)

main = do
    contents <- readFile "input.txt"
    let grid = lines contents
        startX = head [x | (x, c) <- zip [0..] (head grid), c == '|']
        (letters, _, _, _) = followPath grid startX 0 Down [] 0
    putStrLn letters

followPath :: [String] -> Int -> Int -> Direction -> String -> Int -> (String, Int, Int, Direction)
followPath grid x y dir letters steps
    | current == ' ' = (letters, x, y, dir)
    | current `elem` ['|', '-'] || (current >= 'A' && current <= 'Z') = move grid x y dir letters steps
    | current == '+' = turn grid x y dir letters steps
    where current = (grid !! y) !! x

move :: [String] -> Int -> Int -> Direction -> String -> Int -> (String, Int, Int, Direction)
move grid x y dir letters steps
    | dir == Up = followPath grid x (y-1) dir (collectLetter grid x y letters) (steps+1)
    | dir == Down = followPath grid x (y+1) dir (collectLetter grid x y letters) (steps+1)
    | dir == Lft = followPath grid (x-1) y dir (collectLetter grid x y letters) (steps+1)
    | dir == Rght = followPath grid (x+1) y dir (collectLetter grid x y letters) (steps+1)

turn :: [String] -> Int -> Int -> Direction -> String -> Int -> (String, Int, Int, Direction)
turn grid x y dir letters steps
    | dir == Up || dir == Down = if (grid !! y) !! (x-1) == '-' then move grid (x-1) y Lft letters steps else move grid (x+1) y Rght letters steps
    | dir == Lft || dir == Rght = if (grid !! (y-1)) !! x == '|' then move grid x (y-1) Up letters steps else move grid x (y+1) Down letters steps

collectLetter :: [String] -> Int -> Int -> String -> String
collectLetter grid x y letters
    | current >= 'A' && current <= 'Z' = letters ++ [current]
    | otherwise = letters
    where current = (grid !! y) !! x